sys/kern/dev_procstat.c

```diff
@@ -0,0 +1,194 @@
+#define MAX_PROCSTAT 10
+
+static char proc_state[4] = {
+  [PS_NORMAL] = 'R',
+  [PS_STOPPED] = 'S',
+  [PS_DYING] = 'D',
+  [PS_ZOMBIE] = 'Z',
+};
+
+typedef struct ps_entry {
+  pid_t pid;
+  pid_t ppid;
+  uid_t uid;
+  pgid_t pgrp;
+  sid_t sid;
+  proc_state_t proc_state;
+  char *elfpath;
+} ps_entry_t;
+
+typedef struct ps_buf {
+  int nproc;
+  int cur_proc;
+  size_t offset; /* offset in buffer of current proc string */
+  size_t psize;  /* size of current proc string */
+  char buf[MAX_P_STRING];
+  ps_entry_t ps[];
+} ps_buf_t;
+
+static int ps_opencnt = 0;
+static mtx_t procstat_lock;
+
+static int dev_procstat_read(file_t *f, uio_t *uio);
+static int dev_procstat_close(vnode_t *v, file_t *fp);
+static int dev_procstat_open(vnode_t *v, int mode, file_t *fp);
+
+static fileops_t dev_procstat_fileops = {
+  .fo_read = dev_procstat_read,
+  .fo_close = default_vnclose,
+  .fo_stat = default_vnstat,
+  .fo_ioctl = default_vnioctl,
+};
+
+static vnodeops_t dev_procstat_vnodeops = {
+  .v_open = dev_procstat_open,
+  .v_close = dev_procstat_close,
+};
+
+static void ps_entry_fill(ps_entry_t *pe, proc_t *p) {
+  SCOPED_MTX_LOCK(&p->p_lock);
+  pe->uid = p->p_cred.cr_euid;
+  pe->pid = p->p_pid;
+  pe->ppid = p->p_parent->p_pid;
+  pe->pgrp = p->p_pgrp->pg_id;
+  pe->sid = p->p_pgrp->pg_session->s_sid;
+  pe->proc_state = p->p_state;
+  pe->elfpath = kstrndup(M_TEMP, p->p_elfpath, PATH_MAX);
+}
+
+/* buf must be at least MAX_P_STRING long
+ * returns length of written string
+ */
+static int ps_entry_tostring(char *buf, ps_entry_t *pe) {
+  int r = snprintf(buf, MAX_P_STRING, "%d\t%d\t%d\t%d\t%d\t%c\t%s\n", pe->uid,
+                   pe->pid, pe->ppid, pe->pgrp, pe->sid,
+                   proc_state[pe->proc_state], pe->elfpath);
+
+  return MIN(r, MAX_P_STRING);
+}
+
+static int dev_procstat_open(vnode_t *v, int mode, file_t *fp) {
+  WITH_MTX_LOCK (&procstat_lock) {
+    if (ps_opencnt >= MAX_PROCSTAT)
+      return EMFILE;
+    ps_opencnt++;
+    assert(ps_opencnt <= MAX_PROCSTAT);
+  }
+
+  int error;
+  proc_t *p;
+  ps_buf_t *ps;
+
+  ps = kmalloc(M_TEMP, MAX_PROC * sizeof(ps_entry_t) + sizeof(ps_buf_t), 0);
+  if (ps == NULL)
+    return ENOMEM;
+
+  ps->nproc = 0;
+
+  WITH_MTX_LOCK (all_proc_mtx) {
+    TAILQ_FOREACH (p, &proc_list, p_all) {
+      if (p->p_pid == 0)
+        continue; /* we don't want to show proc0 to user */
+      if (ps->nproc >= MAX_PROC)
+        goto out;
+      ps_entry_fill(&ps->ps[ps->nproc++], p);
+    }
+    TAILQ_FOREACH (p, &zombie_list, p_zombie) {
+      if (ps->nproc >= MAX_PROC)
+        goto out;
+      ps_entry_fill(&ps->ps[ps->nproc++], p);
+    }
+  }
+
+out:
+  if ((error = vnode_open_generic(v, mode, fp)))
+    return error;
+
+  /* get first proc into buffer */
+  ps->psize = ps_entry_tostring(ps->buf, &ps->ps[0]);
+  ps->offset = 0;
+  ps->cur_proc = 0;
+
+  fp->f_ops = &dev_procstat_fileops;
+  fp->f_data = ps;
+  return 0;
+}
+
+static int dev_procstat_read(file_t *f, uio_t *uio) {
```

Comment:
Since this is a file operation and not a vnode operation, there's no implicit synchronization between `read()` calls on the same open file. If a process forks while holding an open file descriptor to `/dev/procstat`, you can have two different processes calling `read()` concurrently on the same `file_t`. You need to synchronize them.

Reply:
Added `ps_buf_t::lock` to synchornize reads.